components/webgpu/wgpu_thread.rs

```diff
@@ -868,21 +872,70 @@ impl WGPU {
+                        color_attachments,
+                        depth_stencil_attachment,
+                        device_id: _device_id,
+                    } => {
+                        let global = &self.global;
+                        let desc = &RenderPassDescriptor {
+                            label,
+                            color_attachments: color_attachments.into(),
+                            depth_stencil_attachment: depth_stencil_attachment.as_ref(),
+                            timestamp_writes: None,
+                            occlusion_query_set: None,
+                        };
+                        let (pass, error) = gfx_select!(
+                            command_encoder_id => global.command_encoder_create_render_pass_dyn(
+                                command_encoder_id,
+                                desc,
+                        ));
+                        assert!(
+                            self.render_passes
+                                .insert(render_pass_id, (pass, error.is_none()))
+                                .is_none(),
+                            "RenderPass should not exist yet."
+                        );
+                        // TODO: Command encoder state errors
+                        // self.maybe_dispatch_wgpu_error(device_id, error);
+                    },
+                    WebGPURequest::RenderPassCommand {
+                        render_pass_id,
+                        render_command,
+                        device_id,
+                    } => {
+                        if let Some((pass, valid)) = self.render_passes.get_mut(&render_pass_id) {
+                            *valid &= do_render_command(&self.global, pass, render_command).is_ok();
+                        } else {
+                            self.maybe_dispatch_error(
+                                device_id,
+                                Some(Error::Validation("pass already ended".to_string())),
+                            );
+                        };
+                    },
                     WebGPURequest::EndRenderPass {
-                        render_pass,
+                        render_pass_id,
                         device_id,
+                        command_encoder_id,
                     } => {
-                        if let Some(render_pass) = render_pass {
-                            let command_encoder_id = render_pass.parent_id();
-                            let global = &self.global;
-                            let result = gfx_select!(command_encoder_id => global.render_pass_end(&render_pass));
-                            self.maybe_dispatch_wgpu_error(device_id, result.err())
+                        // TODO: Command encoder state error
+                        if let Some((mut pass, valid)) = self.render_passes.remove(&render_pass_id)
+                        {
+                            if pass.end(&self.global).is_ok() && !valid {
```

Comment:
I think `end` cannot be ok if the pass is not valid, so we can simply check for an error?

Reply:
Actually it can, because wgpu does not store vaildness of passes. This is done the way it is, because we want to do steps 1-4 and only if we reached far enough (end returns ok), we actually check valid (step 5 in [spec](https://www.w3.org/TR/webgpu/#dom-gpurenderpassencoder-end)).

Reply:
Ok, would be good to document then.